Enables Yarn 2

arcanis opened the pull request:
## What I did

Upgrading the Yarn binary to the latest stable release (2.4), which should make them faster.

I've configured the project to stay on the node_modules linker since it makes the upgrade fairly painless. Tests pass locally, I'll see how it goes on CI and tweak the command line references accordingly (things like `--frozen-lockfile` being renamed to `--immutable`, etc).

Notable changes:

- The Yarn binary is checked within this repository. This mean that all your contributors will now be ensured to use the exact same version of Yarn, regardless what they have installed globally.

- The root package.json now lists all workspaces as dependencies (using the `workspace:` protocol). This is because in Yarn 2 the root workspace doesn't implicitly depend on all workspaces anymore, they have to be explicitly listed. In theory the dependencies should be listed in each workspace that depends on them (so for example `ui` should depend on `api`) but I'm worried it would lead to a gigantic diff, so for now the repo keeps relying on the hoisting to work.

## How to test

The existing CI pipeline should be enough to validate that the upgrade is safe. One potential risky part is the publish workflow; I've heard it's already covered by tests using Verdaccio so it should be fine, but if someone can double check it would be perfect!

arcanis committed:
Enables Yarn 2

arcanis committed:
Removes the accidental --version flag

arcanis committed:
Updates circleci

arcanis committed:
Removes the examples from the deps

arcanis committed:
Removes the examples from the deps

arcanis committed:
Tries to preserve the original intent

arcanis commented:
What are those TeamCity tests? Can't see their content 🤔

arcanis committed:
No need for bootstrap

arcanis committed:
Debug

arcanis committed:
Fixes bootstrap script

arcanis committed:
Tries enabling ngrok

arcanis committed:
Disables colors

wKich commented:
@arcanis, you can login as a guest

arcanis committed:
Prints the node version

arcanis committed:
Prints the node version

gaetanmaisse commented:
@arcanis I think you don't need to worry about the TeamCity tests for now, our "official" CI workflows are the ones from CircleCI. Plus, if I remember correctly I had some issue with them in a previous CI PR because there were always using workflow defined in `next` branch and not the one I updated in my PR.

gaetanmaisse commented:
I have just check out your branch locally, ran yarn install + build + test, and everything was ok 💪🏻  

Doing the same in the CI executor (with ssh connection) and it's stuck after the resolution step (in the fetch step? 🤷🏻 ), maybe the [current `medium` executor](https://github.com/storybookjs/storybook/blob/06fba1d51153829a7acd674346dff4a7c47b7125/.circleci/config.yml#L79) isn't powerful enough, you can try with a `large` or `xlarge` one. 

I'm really interested in this PR, feels free to ping me on discord if you want!

arcanis committed:
Let's try large

arcanis committed:
Removes --immutable

arcanis committed:
Runs install before the bootstrap

arcanis committed:
Fixes publish

arcanis committed:
Allows http access to localhost

arcanis committed:
Runs the bootstrap on build

arcanis committed:
Tries large again (solves a TS build?)

gaetanmaisse commented:
Note for myself: As everything is working with Yarn 2 now, the Yarn 2 e2e tests must be reworked and renamed to Yarn PnP

merceyz commented:
The e2e tests should avoid using npx as well, as that wont use the Yarn cache which will slow them down

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
chore: temporary add `eslint-config-airbnb` in base package.json otherwise eslint won't run

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse committed:
e2e: set Yarn 2 linker to node_modules for the "default" e2e tests

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse committed:
ci: update cache key used in CircleCI workflow

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse commented:
Tested locally https://github.com/storybookjs/presets/pull/181 is fixing a lot of the current issues, I can't merge that PR as required CI checks are outdated and I'm not admin on that repo so I can't change them 🙈 . I have asked some people to take a look, hope it will be updated soon, in the meanwhile we are a bit stuck.

gaetanmaisse committed:
chore: bump `@storybook/preset-create-react-app` to latest alpha

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse committed:
fix: add missing dep in some CRA examples

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse committed:
chore: clean usage of `skipLibCheck` and `skipDefaultLibCheck` TS option

`skipDefaultLibCheck` is deprecated so there is no need to add it
anymore.

I added `skipLibCheck` in the angular example because it looks like
there are some type incompatibility between some types coming from
node_modules we have in the monorepo. (Not really a surprise with the
number of deps we have)

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

gaetanmaisse committed:
chore: explicitly call `prebuild-storybook` scripts

Signed-off-by: Gaëtan Maisse <gaetanmaisse@gmail.com>

merceyz commented on examples/angular-cli/package.json:
```diff
@@ -5,8 +5,7 @@
     "build": "ng build",
-    "prebuild-storybook": "yarn storybook:prebuild",
-    "build-storybook": "build-storybook -s src/assets",
+    "build-storybook": "yarn storybook:prebuild && build-storybook -s src/assets",
```
Do note that script names containing `:` (colon) are global and thus don't allow duplicates
https://yarnpkg.com/getting-started/qa/#how-to-share-scripts-between-workspaces

gaetanmaisse replied:
Good to know! I think we will need to do a big clean around that then.

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
fix: add missing dep in storyshot and set the same version of @angular everywhere

gaetanmaisse committed:
ci: install all dependencies in `build` step to avoid hoisting issue

gaetanmaisse committed:
chore: simplify `yarn test` script by using Jest directly instead of a custom CLI

For now, there is still a custom node script for a few weeks to still handle `yarn test --core` command.
The `--puppeteer` option is now a standalone NPM script: `yarn test:puppeteer`.

I also updated the CI jobs and the doc.

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
ci: remove `-w 2` option in unit tests job of CircleCI workflow

gaetanmaisse committed:
chore: reduce verdaccio startup time tiemout to 20s

gaetanmaisse committed:
chore: call directly Jest for `test` script instead of using a custom command

Also add `runInBand` and `ci` Jest options when running on CI.

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
chore: fix unit test related to WC in addons-docs

gaetanmaisse committed:
ci: fix yarn npm registry config

gaetanmaisse committed:
chore: fix lint error

gaetanmaisse committed:
ci: add `--immutable` in yarn install

gaetanmaisse committed:
fix(cli): do not reinstall packages that were already there as deps or devDeps

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
e2e: remove `-W` option of Yarn command

gaetanmaisse committed:
Merge branch 'next' into mael/storybook

gaetanmaisse committed:
test: update `core-presets` snapshots to match core js version

gaetanmaisse committed:
e2e: remove all exotic options send to yarn commands

gaetanmaisse commented:
Closing in favour of smaller and more focused PRs:
 - https://github.com/storybookjs/storybook/pull/13871
 - https://github.com/storybookjs/storybook/pull/13872
 - https://github.com/storybookjs/storybook/pull/13876
 - https://github.com/storybookjs/storybook/pull/13878
 - https://github.com/storybookjs/storybook/pull/13907 <-- only this one is really related to Yarn 2 migration
